src/pcre2_match.c

```diff
@@ -1929,6 +1929,104 @@ fprintf(stderr, "++ %2ld op=%3d %s\n", Fecode - mb->start_code, *Fecode,
 #undef Loc
 
 
+    /* ===================================================================== */
+    /* Match a complex, set-based character class. These opcodes are used when
+    there is complex nesting or logical operations within the character
+    class. */
+
+    case OP_ECLASS:
+      {
+      uint32_t stack = 0;
+      PCRE2_SPTR ecode_end = Fecode + GET(Fecode, 1);
+
+      Fecode += 1 + LINK_SIZE;  /* Advance past OP_ECLASS header + length. */
+
+      /* Read in the character to test. */
+
+#ifdef SUPPORT_UNICODE
+      if (utf)
+        {
+        if (Feptr >= mb->end_subject)
+          {
+          SCHECK_PARTIAL();
+          RRETURN(MATCH_NOMATCH);
+          }
+        GETCHARINC(fc, Feptr);
+        }
+      else
+#endif
+      /* Not UTF mode */
+        {
+          if (Feptr >= mb->end_subject)
+            {
+            SCHECK_PARTIAL();
+            RRETURN(MATCH_NOMATCH);
+            }
+          fc = *Feptr++;
+        }
+
+      /* Now do a little loop, until we reach the end of the ECLASS. */
+      while (Fecode < ecode_end)
+        {
+        switch (*Fecode)
+          {
+          case OP_ECLASS_OR:
+          ++Fecode;
+          stack = (stack >> 1) | (stack & 1u);
+          break;
+
+          case OP_ECLASS_AND:
+          ++Fecode;
+          stack = (stack >> 1) & (stack & 1u);
```

Comment:
Note: this clears all upper bits of the stack. `(stack >> 1) & (stack | ~(uint32_t)1u)` is the correct form. Sub is also a similar case.

Reply:
Ouch, that was embarrassing, thank you.

I've started doing more testing, and fixed a few silly bugs.